ddb-popper/ddb-popper.js

```diff
@@ -0,0 +1,200 @@
+            windowFeatures: "resizeable,scrollbars,location=no,width=768,height=968"
+        }  
+    }
+
+    /**
+     * Metadata used to register settings
+     * @returns {Object} Settings Metadata
+     */
+    static get SETTINGS_META() {
+        return {
+            name: "ddbCharacterURL",
+            type: String,
+            scope: "world",
+            default: ""
+        }
+    }
+
+    /**
+     * Hooks on render of ActorSheet5eCharacter in order to insert the DDB Button
+     */
+    _hookRenderActorSheet() {
+        Hooks.on("renderActorSheet5eCharacter", (app, html, data) => {
+            this._addDDBButton(html, data);
+        });
+    }
+
+    /**
+     * For a given actor's data, get the D&D Beyond URL if it exists in settings, 
+     * and if not, register and try again
+     * @param {Object} data -- an actor's data property
+     * @returns {String} ddbURL -- the matching D&D Beyond URL (if any)
+     */
+    async _getActorDDBURL(data) {
+        let ddbURL;
+
+        try {
+            ddbURL = await game.settings.get(DDBPopper.CONFIG.moduleName, data.actor._id);
+        } catch (e) {
+            if(e.message == "This is not a registered game setting") {
+                await game.settings.register(DDBPopper.CONFIG.moduleName, data.actor._id, DDBPopper.SETTINGS_META);
+                ddbURL = await game.settings.get(DDBPopper.CONFIG.moduleName, data.actor._id);
+            } else {
+                ddbURL = null;
+                throw(e);
+            }
+        } finally {
+            return ddbURL;
+        }
+    }
+
+    /**
+     * Adds the DDB button to the supplied html
+     * @param {Object} html -- the html of the ActorSheet5eCharacter
+     * @param {Object} data -- the data of the actor
+     */
+    async _addDDBButton (html, data) {
+        /**
+         * Finds the header and the close button
+         */
+        const windowHeader = html.parent().parent().find(".window-header");
+        const windowCloseBtn = windowHeader.find(".close");
+
+        /**
+         * jquery reference to the D&D Beyond button to add to the sheet
+         */
+        const ddbButton = $(
+            `<a class="ddb-popup">
+                <img src=${DDBPopper.CONFIG.ddbLogo} style=${DDBPopper.CONFIG.imgStyle}> DDB
+            </a>`
+        );
+        
+        /**
+         * Remove any existing instances of the button on the sheet
+         */
+        windowHeader.find('.ddb-popup').remove();
+        windowCloseBtn.before(ddbButton);
+
+        /**
+         * When the DDB button is clicked, lookup the DDB URL, then look for an existing popup,
+         * and if none exists, and the DDB URL is null, open the form to set the URL, 
+         * otherwise if the popup does not exist but the DDB URL is not null, create the popup
+         * or else simply focus the existing popup
+         */
+        ddbButton.click(async ev => {
+            ev.preventDefault();
```

Comment:
What's the reason for this? (May just want a comment on why it's required)

Reply:
This is a standard method for the `event` interface per MDN: https://developer.mozilla.org/en-US/docs/Web/API/Event/preventDefault it's seemingly standard practice to call this on any non-standard event calls (eg. overwriting context click etc)

Reply:
Yeah, but you usually only do this if it's causing some issue by emitting the event as is. Harmless, but just extra code that I don't see the benefit of.

Reply:
For the onClick probably not needed. The context one though is probably necessary to prevent accidental fires of the context menu (I believe it's blocked upstream anyway though).